librarymanagement/src/main/scala/sbt/internal/librarymanagement/ivyint/SbtChainResolver.scala

```diff
@@ -71,188 +72,235 @@ private[sbt] case class SbtChainResolver(
+        // If `None` do nothing -- modules without artifacts. Otherwise cache.
+        artifactOpt.foreach { artifactRef =>
+          val dep = toSystem(descriptor)
+          val first = toSystem(resolvedDescriptor).getAllArtifacts.head
+          val options = getCacheOptions(data)
+          val cacheManager = getRepositoryCacheManager
+          cacheManager.cacheModuleDescriptor(resolver, artifactRef, dep, first, null, options)
         }
-      case None if errors.size > 1 =>
-        val err =
-          (errors.toList map { IvyStringUtils.getErrorMessage }).mkString("\n\t", "\n\t", "\n")
-        throw new RuntimeException(s"several problems occurred while resolving $dd:$err")
-      case _ =>
-        if (resolved == mrOpt) resolved.orNull
-        else (mrOpt map { resolvedRevision }).orNull
+        resolvedModule
+      }
+    }
+
+    def resolveByAllMeans(
+        cachedModule: Option[ResolvedModuleRevision],
+        useLatest: Boolean,
+        interResolver: Option[DependencyResolver],
+        resolveModules: () => Seq[Either[Throwable, TriedResolution]],
+        dd: DependencyDescriptor,
+        data: ResolveData
+    ): Option[ResolvedModuleRevision] = {
+      val internallyResolved: Option[ResolvedModuleRevision] = {
+        if (!updateOptions.interProjectFirst) None
+        else interResolver.flatMap(resolver => Option(resolver.getDependency(dd, data)))
+      }
+      val internalOrExternal = internallyResolved.orElse {
+        val foundRevisions: Seq[(ResolvedModuleRevision, DependencyResolver)] =
+          resolveModules().collect { case Right(Some(x)) => x }
+        if (useLatest) resolveLatest(foundRevisions, dd, data)
+        else foundRevisions.reverse.headOption.map(_._1) // Resolvers are hit in reverse order
+      }
+      internalOrExternal.orElse(cachedModule)
+    }
+
+    // The ivy implementation guarantees that all resolvers implement `DependencyResolver`
+    def getDependencyResolvers: Vector[DependencyResolver] =
+      getResolvers.toArray.collect { case r: DependencyResolver => r }.toVector
+
+    def findInterProjectResolver(resolvers: Seq[DependencyResolver]): Option[DependencyResolver] =
+      resolvers.find(_.getName == ProjectResolver.InterProject)
+
+    /** Gets the dependency for a given descriptor with the pertinent resolve data.
+     *
+     * This is a custom sbt chain operation that produces better error output and deals with
+     * cases that the conventional ivy resolver does not. It accumulates the resolution of
+     * several resolvers and returns the module which fits the provided resolution strategy.
+     *
+     * These are the differences with regard to the default ivy [[ChainResolver]]:
+     *   1. It skips resolution if "return first" is set to true.
+     *   2. It skips resolution if a previously resolved or cached resolution is found.
+     *   3. It always checks all the resolvers and compares timestamps for changing dependencies.
```

Comment:
Technically, it will traverse all resolvers only if `updateOptions.latestSnapshots` is `true` (by default it's set to `true` I think).

The original comment that I wrote in 8e682691c2bc820640fd14ceea14b33779a2351b says `"regardless of what the "latest strategy" is set, and look for the published date or the module descriptor to sort them."` This is reference to Ivy's [latest-strategies](http://ant.apache.org/ivy/history/2.3.0/settings/latest-strategies.html) setting, which by default is set to **latest-revision**. It might be useful for future developer if we link to sbt/sbt#1520.

This also points to the tradeoff of correctness vs time when using `-SNAPSHOT`.